src/yetibot/commands/aws/formatters.clj

```diff
@@ -105,79 +109,122 @@
   {:result/data  {:user-name UserName, :access-key-id AccessKeyId, :status Status, :secret-access-key SecretAccessKey, :create-date CreateDate}
    :result/value (format "An access key for user %s has been successfully created on %s\nAWS Access Key ID : %s\nAWS Secret Access Key : %s\nStatus : %s\n"
                          UserName CreateDate AccessKeyId SecretAccessKey Status)})
 
-(defmethod format-response ::IAMListAccessKeysResponseReceived
+(defmethod format-iam-response ::IAMListAccessKeysResponseReceived
   [{:keys [AccessKeyMetadata]}]
   {:result/data  {:access-key-metadata AccessKeyMetadata}
    :result/value (map #(format "Access key ID : %s\nStatus : %s\nCreated on : %s\n"
                                (:AccessKeyId %) (:Status %) (:CreateDate %))
                       AccessKeyMetadata)})
 
-(defmethod format-response ::IAMAccessKeyDeleted
+(defmethod format-iam-response ::IAMAccessKeyDeleted
   [response]
   (let [request-id (get-in response [:DeleteAccessKeyResponse :ResponseMetadata :RequestId])]
     {:result/data  {:request-id request-id}
      :result/value (format "Access Key successfully deleted [RequestId=%s]"
-                           request-id)}))
\ No newline at end of file
+                           request-id)}))
+
+;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
+
+(defmulti format-s3-response
+          "Returns a dispatch-value matching the S3 operation that has been successfully invoked
+          or has failed"
+          (fn [response]
+            (let [aws-type (:aws/type (meta response))]
+              (cond
+                (and (s/valid? (s3-response-spec :CreateBucket) response)
+                     (= aws-type :aws.type/CreateBucket)
+                     (not (contains? response :Error))) ::S3BucketCreated
+                (and (s/valid? (s3-response-spec :ListBuckets) response)
+                     (= aws-type :aws.type/ListBuckets)
+                     (not (contains? response :Error))) ::S3BucketListed
+                (and (s/valid? (s3-response-spec :ListObjectsV2) response)
+                     (= aws-type :aws.type/ListObjects)
+                     (not (contains? response :Error))) ::S3ObjectsListed
+                :else ::error))))
+
+(defmethod format-s3-response ::error
+  [response]
+  {:result/error (get-in response [:Error :Message])})
+
+(defmethod format-s3-response ::S3BucketCreated
+  [{:keys [Location]}]
+  {:result/data  {:location Location}
+   :result/value (format "S3 bucket successfully created at %s" Location)})
+
+(defmethod format-s3-response ::S3BucketListed
+  [{:keys [Buckets]}]
+  {:result/data  {:buckets Buckets}
+   :result/value (map
+                   #(format "Bucket : %s - Created on %s" (:Name %) (:CreationDate %))
+                   Buckets)})
+
+(defmethod format-s3-response ::S3ObjectsListed
+  [{:keys [Contents]}]
+  {:result/data  {:contents Contents}
```

Comment:
Same - more data?

Reply:
Same for the _objects listing_, we have more infos returned by the API. I will paste the _format_ here later and we will decide which infos we will keep in the formatted response.

Reply:
for the `:ListObjectsV2` API call, we have the following _spec_ : 

```
:ListObjectsV2 {:name "ListObjectsV2",
                 :documentation "<p>Returns some or all (up to 1000) of the objects in a bucket. You can use the request parameters as selection criteria to return a subset of the objects in a bucket. Note: ListObjectsV2 is the revised List Objects API and we recommend you use this revised API for new application development.</p>",
                 :request {:Prefix string,
                           :StartAfter string,
                           :Bucket string,
                           :EncodingType string,
                           :Delimiter string,
                           :FetchOwner boolean,
                           :RequestPayer string,
                           :ContinuationToken string,
                           :MaxKeys integer},
                 :required [:Bucket],
                 :response {:Prefix string,
                            :StartAfter string,
                            :EncodingType string,
                            :Delimiter string,
                            :NextContinuationToken string,
                            :CommonPrefixes [:seq-of {:Prefix string}],
                            :ContinuationToken string,
                            :Contents [:seq-of
                                       {:Key string,
                                        :LastModified timestamp,
                                        :ETag string,
                                        :Size integer,
                                        :StorageClass string,
                                        :Owner {:DisplayName string, :ID string}}],
                            :MaxKeys integer,
                            :IsTruncated boolean,
                            :Name string,
                            :KeyCount integer}},
```

I only took the `:Contents` but the `:response` key is assoc-ed to much more information, mostly meta-data such as _pagination-related_ stuff (I will also have to think about how to manage it too by the way ^^).
Which other keys do you think we should include in the response for this one ?

Reply:
I think everything. I'm aiming to make data behind pipes super powerful. I'd prob only omit something if it was massive, unrelated, or a security issue.